### PR TITLE
Stack allocate generic and resilient values

### PR DESCRIPTION
**Pull request by aschwaighofer**
IRGen: Allocate generic/resilient values on the stack instead of on the heap
---------------------------------------------------------------------------
Allocate buffers for local generic/resilient values on the stack. alloc_stack
instructions in the entry block are translated using a dynamic alloca
instruction with variables size. All other alloc_stack instructions in addition
use llvm's stacksave/restore instrinsics to reset the stack (they could be
executed multiple times and with varying sizes).

Add support of a IRGen lowering SIL pipeline
---------------------------------------------
This pipeline is run as part of IRGen and has access to the IRGenModule.

Passes that run as part of this pipeline can query for the IRGenModule.

We will use it for the AllocStackHoisting pass. It wants to know when a type is of
non-fixed size.

To break the cyclic dependency between IRGen -> SILOptimizer -> IRGen that would
arise from the SILPassManager having to know about the createIRGENPASS()
function IRGen passes instead of exposing this function dynamically have to add
themselves to the pass manager.

Add an alloc_stack hoisting pass
---------------------------------
Hoist alloc_stack instructions of 'generic' or resilient type to the entry
block.  Ad the same time also perform a very simple stack coloring analysis.
This does not use a true liveness-analysis yet but rather employs some simple
conservative checks to see whether the live ranges of two alloc_stacks might
interfere.

AllocStackHoisting is an IRGen SIL pass. This allows for using IRGen's type
lowering information.

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
Awesome!

**Comment by slavapestov** [comment]
Would it be better to extend SIL type lowering to distinguish fixed size address only types (weak references, existentials) from the generic/resilient case, eliminating the weird dependency on IRGen in the new SIL pass?

File: include/swift/SIL/SILFunction.h [hunk 1]
```diff
@@ -618,6 +618,9 @@ class SILFunction
   SILBasicBlock &front() { return *begin(); }
   const SILBasicBlock &front() const { return *begin(); }
 
+  SILBasicBlock *entryBB() { return &front(); }
```
Comment (gottesmm) [inline include/swift/SIL/SILFunction.h]: Can you change this to getEntryBlock()?

File: tools/sil-opt/SILOpt.cpp [hunk 2]
```diff
@@ -318,7 +330,12 @@ int main(int argc, char **argv) {
   } else if (OptimizationGroup == OptGroup::Performance) {
     runSILOptimizationPasses(*CI.getSILModule());
   } else {
-    runCommandLineSelectedPasses(CI.getSILModule());
+    auto *SILMod = CI.getSILModule();
+    {
+      auto T = irgen::createIRGenModule(SILMod, getGlobalLLVMContext());
+      runCommandLineSelectedPasses(SILMod, T.second);
+      irgen::deleteIRGenModule(T);
+    }
```
Comment (gottesmm) [inline tools/sil-opt/SILOpt.cpp]: Question. I am worried about being able to mix IRGen and Performance passes. What is to stop me from doing so and if I do so what will happen? Should we have an additional SIL stage or something like that? (Just thinking out loud).

One could even argue that perhaps we need a separate entry point (like autolink is to the driver that only can optimize lowered SIL) or something like that.

Like I said I am just thinking out loud (unsure).

**Review by gottesmm — COMMENTED** [review]
A question + a rename request.

**Comment by gottesmm** [comment]
Also, why couldn't we make these passes in a separate library. Then you don't need to do the dynamic stuff. libswiftSILIRGenPasses...

**Comment by gottesmm** [comment]
(Even though I have a couple of high level thoughts... +1! Overall!)

File: tools/sil-opt/SILOpt.cpp (hunk 2 above)
Comment (aschwaighofer) [inline tools/sil-opt/SILOpt.cpp]: At this point nothing bad will happen and you are free to do so

Comment (gottesmm) [inline tools/sil-opt/SILOpt.cpp]: Ok. So what _will_ happen now?

Comment (gottesmm) [inline tools/sil-opt/SILOpt.cpp]: Ah ok. The pass was hidden by github's interface. You are just using the size information, nothing else. My worries are gone!

**Comment by aschwaighofer** [comment]
@slavapestov  I don't see a good reason to replicate the recursive type lowering at the SIL level?

**Comment by aschwaighofer** [comment]
@slavapestov Once we implement a SIL lowering pass that keeps large structs in memory (instead of aggressively exploding it) this pass will also very likely want to query IRGen for lowering properties such as size and will run in this IRGen pipeline. So will Andy's work (although it is not dependent on having an IRGenModule AFAIK).

@gottesmm The pass manager has a dependency on the createPASSName() function.

Moving the pass into a separate library does not fix that.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Well, SIL already does a recursive walk over value types when lowering, it just computes less stuff than IRGen.

But as long as you run the IRGen passes after serialization, it should be ok.

One day we might want to come up with some abstractions to factor out duplication between SIL and IRGen type lowering. Sema has to do a recursive walk over value types too, to diagnose infinite types.

@jckarter Any thoughts?

**Comment by jckarter** [comment]
SIL type lowering could definitely track some more interesting properties such as "empty", "uninhabited", and "fixed-size" that can be figured out irrespective of the concrete IR layout. I think it still makes sense to have a transitional stage between SIL and IRGen where we do SIL transformations based on properties that normally aren't determined till IRGen.

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by aschwaighofer** [comment]
@swift-ci Please benchmark

**Comment by aschwaighofer** [comment]
@swift-ci Please benchmark

**Pull request merged**